src/coreclr/jit/importer.cpp

```diff
@@ -9761,9 +9767,28 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
                 {
                     // Null check is sometimes needed for ready to run to handle
                     // non-virtual <-> virtual changes between versions
+                    bool addNullcheck = true;
                     if (callInfo->nullInstanceCheck)
                     {
-                        call->gtFlags |= GTF_CALL_NULLCHECK;
+                        GenTree* thisArg = impStackTop(sig->numArgs).val;
+                        if (thisArg->OperIs(GT_RET_EXPR))
+                        {
+                            GenTreeCall* thisAsCall = thisArg->AsRetExpr()->gtInlineCandidate->AsCall();
+                            if (thisAsCall->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC)
+                            {
+                                NamedIntrinsic ni = lookupNamedIntrinsic(thisAsCall->gtCallMethHnd);
+                                if ((ni == NI_System_Collections_Generic_EqualityComparer_get_Default) ||
+                                    (ni == NI_System_Collections_Generic_Comparer_get_Default))
+                                {
```

Comment:
Would be nice to re-use the pattern in `fgIsCurrentCallDceCandidate` here -- maybe generalize that to take a method handle?

Reply:
Technically this is a bit different logic, it's a quick check that we can skip a nullcheck, we can add here more intrinsics, potentially:

NI_System_Threading_Thread_get_CurrentThread
NI_System_Array_Clone
NI_System_Object_MemberwiseClone
NI_System_Object_GetType

all of these are not expected to return null afair